# <center>Data Mining Project</center>

<center>
Master in Data Science and Advanced Analytics <br>
NOVA Information Management School
</center>

** **
## <center>*ABCDEats Inc*</center>

<center>
Group 19 <br>
Jan-Louis Schneider, 20240506  <br>
Marta Boavida, 20240519  <br>
Matilde Miguel, 20240549  <br>
Sofia Gomes, 20240848  <br>
</center>

** **


In [1]:
import pandas as pd 
import numpy as np
import scipy

import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import chi2_contingency
import scipy.stats as stats
import warnings

from math import ceil
from sklearn.impute import KNNImputer

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

In [2]:
df = pd.read_csv("../dataset/DM2425_ABCDEats_DATASET.csv")

## <span style="color:salmon"> <center>Explore Data</center> </span> 

### <span style="color:salmon"> 1. Explore Data Analysis </span> 
Conduct an in-depth exploration of the dataset. Summarise key statistics for the data, and discuss their possible implications.

#### <span style="color:salmon"> 1.1 Explore Dataset </span>  
In order to a better understanding the dataset, we used a funcion head, to see the first ten lines.

In [3]:
df.head(10)

customer_id customer_region  customer_age  vendor_count  product_count  \
0  1b8f824d5e            2360          18.0             2              5   
1  5d272b9dcb            8670          17.0             2              2   
2  f6d1b2ba63            4660          38.0             1              2   
3  180c632ed8            4660           NaN             2              3   
4  4eb37a6705            4660          20.0             2              5   
5  6aef2b6726            8670          40.0             2              2   
6  8475ee66ef            2440          24.0             2              2   
7  f2f53bcc67            8670          27.0             2              3   
8  5b650c89cc            2360          20.0             3              4   
9  84775a7237            8670          20.0             2              3   

   is_chain  first_order  last_order last_promo payment_method  CUI_American  \
0         1          0.0           1   DELIVERY           DIGI          0.00   
1         2          0.0           1   DISCOUNT           DIGI         12.82   
2         2          0.0           1   DISCOUNT           CASH          9.20   
3         1          0.0           2   DELIVERY           DIGI          0.00   
4         0          0.0           2          -           DIGI         14.57   
5         0          0.0           2    FREEBIE           DIGI          0.00   
6         2          0.0           2          -           CARD          5.88   
7         2          0.0           2   DISCOUNT           DIGI         11.71   
8         2          0.0           3   DISCOUNT           DIGI          2.75   
9         0          0.0           3   DELIVERY           CARD          0.00   

   CUI_Asian  CUI_Beverages  CUI_Cafe  CUI_Chicken Dishes  CUI_Chinese  \
0       0.00           0.00       0.0                 0.0          0.0   
1       6.39           0.00       0.0                 0.0          0.0   
2       0.00           0.00       0.0                 0.0          0.0   
3      13.70           0.00       0.0                 0.0          0.0   
4      40.87           0.00       0.0                 0.0          0.0   
5      24.92           0.00       0.0                 0.0          0.0   
6       0.00           1.53       0.0                 0.0          0.0   
7       0.00          24.40       0.0                 0.0          0.0   
8       0.00           0.00       0.0                 0.0          0.0   
9      32.48           0.00       0.0                 0.0          0.0   

   CUI_Desserts  CUI_Healthy  CUI_Indian  CUI_Italian  CUI_Japanese  \
0           0.0         0.00       28.88          0.0           0.0   
1           0.0         0.00        0.00          0.0           0.0   
2           0.0         0.00        0.00          0.0           0.0   
3           0.0         0.00       17.86          0.0           0.0   
4           0.0         0.00        0.00          0.0           0.0   
5           0.0         0.00        0.00          0.0           0.0   
6           0.0         0.00        0.00          0.0           0.0   
7           0.0         0.00        0.00          0.0           0.0   
8           0.0         4.39        0.00          0.0           0.0   
9           0.0         0.00        0.00          0.0           0.0   

   CUI_Noodle Dishes  CUI_OTHER  CUI_Street Food / Snacks  CUI_Thai  DOW_0  \
0                0.0        0.0                       0.0       0.0      1   
1                0.0        0.0                       0.0       0.0      1   
2                0.0        0.0                       0.0       0.0      1   
3                0.0        0.0                       0.0       0.0      0   
4                0.0        0.0                       0.0       0.0      0   
5                0.0        0.0                       0.0       0.0      0   
6                0.0        0.0                       0.0       0.0      0   
7                0.0        0.0                       0.0      

Then, we print all the columns of the dataset.

In [4]:
df.columns.values

array(['customer_id', 'customer_region', 'customer_age', 'vendor_count',
       'product_count', 'is_chain', 'first_order', 'last_order',
       'last_promo', 'payment_method', 'CUI_American', 'CUI_Asian',
       'CUI_Beverages', 'CUI_Cafe', 'CUI_Chicken Dishes', 'CUI_Chinese',
       'CUI_Desserts', 'CUI_Healthy', 'CUI_Indian', 'CUI_Italian',
       'CUI_Japanese', 'CUI_Noodle Dishes', 'CUI_OTHER',
       'CUI_Street Food / Snacks', 'CUI_Thai', 'DOW_0', 'DOW_1', 'DOW_2',
       'DOW_3', 'DOW_4', 'DOW_5', 'DOW_6', 'HR_0', 'HR_1', 'HR_2', 'HR_3',
       'HR_4', 'HR_5', 'HR_6', 'HR_7', 'HR_8', 'HR_9', 'HR_10', 'HR_11',
       'HR_12', 'HR_13', 'HR_14', 'HR_15', 'HR_16', 'HR_17', 'HR_18',
       'HR_19', 'HR_20', 'HR_21', 'HR_22', 'HR_23'], dtype=object)

The function info() allow us to get information about the index dtype and columns, non-null values and memory usage.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31888 entries, 0 to 31887
Data columns (total 56 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   customer_id               31888 non-null  object 
 1   customer_region           31888 non-null  object 
 2   customer_age              31161 non-null  float64
 3   vendor_count              31888 non-null  int64  
 4   product_count             31888 non-null  int64  
 5   is_chain                  31888 non-null  int64  
 6   first_order               31782 non-null  float64
 7   last_order                31888 non-null  int64  
 8   last_promo                31888 non-null  object 
 9   payment_method            31888 non-null  object 
 10  CUI_American              31888 non-null  float64
 11  CUI_Asian                 31888 non-null  float64
 12  CUI_Beverages             31888 non-null  float64
 13  CUI_Cafe                  31888 non-null  float64
 14  CUI_Ch

#### <span style="color:salmon"> 1.2 Incorrect Data Types </span> 
The function dtypes show the types of the columns.

In [6]:
df.dtypes

customer_id                  object
customer_region              object
customer_age                float64
vendor_count                  int64
product_count                 int64
is_chain                      int64
first_order                 float64
last_order                    int64
last_promo                   object
payment_method               object
CUI_American                float64
CUI_Asian                   float64
CUI_Beverages               float64
CUI_Cafe                    float64
CUI_Chicken Dishes          float64
CUI_Chinese                 float64
CUI_Desserts                float64
CUI_Healthy                 float64
CUI_Indian                  float64
CUI_Italian                 float64
CUI_Japanese                float64
CUI_Noodle Dishes           float64
CUI_OTHER                   float64
CUI_Street Food / Snacks    float64
CUI_Thai                    float64
DOW_0                         int64
DOW_1                         int64
DOW_2                       

After analyse the datatype, we realise that there are some columns that have an incorrect type, like customer_age, first_order and HR_0.

In [7]:
# convert "customer_age" from float to int
df["customer_age"]= df["customer_age"].astype("Int64")

In [8]:
# convert "first_order" from float to int
df["first_order"] = df["first_order"].astype("Int64")

In [9]:
# convert "HR_0" from float to int
df["HR_0"] = df["HR_0"].astype("Int64")

Another detail about incorrect data is between product_count and vendor_count. There is information about products that were not purchased but there is information about sellers who sold them.

In [10]:
# Product_count = 0, some have vendor_count = 1. 
product_count_zero = vendor_count_zero = df.loc[df["product_count"]==0, ["customer_id", "vendor_count", "product_count", "first_order", "last_order"]]
print(f"Total number of customers with product_count equal to 0: {len(product_count_zero)} customers")

Total number of customers with product_count equal to 0: 156 customers


In [11]:
# unlock vendor count = 0
vendor_count_zero = df.loc[df["vendor_count"]==0, ["customer_id", "vendor_count", "product_count", "first_order", "last_order"]]
print(f"Total number of customers with vendor_count and product_count equal to 0: {len(vendor_count_zero)} customers")

Total number of customers with vendor_count and product_count equal to 0: 138 customers


In [12]:
product_zero_vendor_one = df.loc[(df["product_count"]==0) & (df["vendor_count"]>=1), ["customer_id", "vendor_count", "product_count", "first_order", "last_order"]]
product_zero_vendor_one_percentage = len(product_zero_vendor_one) / len(df) * 100
print(f"Number of customers with product_count = 0 and vendor_count >= 1: {len(product_zero_vendor_one)} customers")
print(f"Percentage of customers with product_count = 0 and vendor_count >= 1: {round(product_zero_vendor_one_percentage, 4)} %")

Number of customers with product_count = 0 and vendor_count >= 1: 18 customers
Percentage of customers with product_count = 0 and vendor_count >= 1: 0.0564 %


#### <span style="color:salmon"> 1.2 Duplicates </span> 
To have a correct data, we need to check if the dataset have duplicates 

In [13]:
# Identify number of duplicates in the dataset.
dup = df["is_chain"].duplicated().value_counts()
dup

is_chain
True     31828
False       60
Name: count, dtype: int64

To see the percentage of duplicates that exists: 

In [14]:
# Percentage of Duplicates in the dataset.
percentage_duplicates = ((df.duplicated().mean()) * 100).round(3)
print(f"Percentage of Duplicates in the dataset is: {percentage_duplicates}%")

Percentage of Duplicates in the dataset is: 0.041%


Then, we removed the duplicates, using the function drop_duplicates().

In [15]:
# Remove duplicates.
df.drop_duplicates(inplace=True)

And confirm if that data have been removed:

In [16]:
# Confirm duplicates were correctly removed
duplicates =df.duplicated().value_counts()
duplicates

False    31875
Name: count, dtype: int64

#### <span style="color:salmon"> 1.3 Missing Values </span> 
Some values are presented like "", so we replaced them with nan values.

In [17]:
df.replace("", np.nan, inplace=True)
df.replace("-", np.nan, inplace=True)
# replace missing values with nan

To see which columns have nan values and how many exists, we used the function isna().sum()

In [18]:
# Check for missing values
missing_values = df.isna().sum()
missing_values

customer_id                     0
customer_region               442
customer_age                  727
vendor_count                    0
product_count                   0
is_chain                        0
first_order                   106
last_order                      0
last_promo                  16744
payment_method                  0
CUI_American                    0
CUI_Asian                       0
CUI_Beverages                   0
CUI_Cafe                        0
CUI_Chicken Dishes              0
CUI_Chinese                     0
CUI_Desserts                    0
CUI_Healthy                     0
CUI_Indian                      0
CUI_Italian                     0
CUI_Japanese                    0
CUI_Noodle Dishes               0
CUI_OTHER                       0
CUI_Street Food / Snacks        0
CUI_Thai                        0
DOW_0                           0
DOW_1                           0
DOW_2                           0
DOW_3                           0
DOW_4         

The variables that have missing values are: customer_region, customer_age, first_order, last_promo and HR_0. With last_promo having a very high percentage (52.5%).

In [19]:
# Percentage of missing values in each variable:
missing_values_customer_region = ((df["customer_region"].isna().mean()) * 100).round(3)
missing_values_customer_age = ((df["customer_age"].isna().mean()) * 100).round(3)
missing_values_first_order = ((df["first_order"].isna().mean()) * 100).round(3)
missing_values_last_promo = ((df["last_promo"].isna().mean()) * 100).round(3)
missing_values_HR_0 = ((df["HR_0"].isna().mean()) * 100).round(3)

print(f"Percentage of missing values in customer_region is: {missing_values_customer_region}%")
print(f"Percentage of missing values in customer_age is: {missing_values_customer_age}%")
print(f"Percentage of missing values in first_order is: {missing_values_first_order}%")
print(f"Percentage of missing values in last_promo is: {missing_values_last_promo}%")
print(f"Percentage of missing values in HR_0 is: {missing_values_HR_0}%")

Percentage of missing values in customer_region is: 1.387%
Percentage of missing values in customer_age is: 2.281%
Percentage of missing values in first_order is: 0.333%
Percentage of missing values in last_promo is: 52.53%
Percentage of missing values in HR_0 is: 3.652%


#### <span style="color:salmon"> 1.4 Numerical variables </span>
Numerical variables represent measurable quantities and can be analyzed mathematically.

In [20]:
# Describe all the columns that have numerical values
df.describe().T

count       mean        std   min   25%   50%  \
customer_age              31148.0  27.507545   7.161828  15.0  23.0  26.0   
vendor_count              31875.0   3.102714   2.771753   0.0   1.0   2.0   
product_count             31875.0   5.668424   6.957933   0.0   2.0   3.0   
is_chain                  31875.0   2.819357   3.977977   0.0   1.0   2.0   
first_order               31769.0  28.469924  24.104626   0.0   7.0  22.0   
last_order                31875.0  63.672376  23.227992   0.0  49.0  70.0   
CUI_American              31875.0   4.877205  11.647043   0.0   0.0   0.0   
CUI_Asian                 31875.0   9.955306  23.561492   0.0   0.0   0.0   
CUI_Beverages             31875.0   2.298224   8.475868   0.0   0.0   0.0   
CUI_Cafe                  31875.0    0.80149   6.428422   0.0   0.0   0.0   
CUI_Chicken Dishes        31875.0   0.768409   3.657986   0.0   0.0   0.0   
CUI_Chinese               31875.0   1.431801   8.193374   0.0   0.0   0.0   
CUI_Desserts              31875.0   0.884719    5.26091   0.0   0.0   0.0   
CUI_Healthy               31875.0    0.95059   5.831747   0.0   0.0   0.0   
CUI_Indian                31875.0   1.631818   7.444679   0.0   0.0   0.0   
CUI_Italian               31875.0   3.234729  11.250094   0.0   0.0   0.0   
CUI_Japanese              31875.0   2.996031  10.182406   0.0   0.0   0.0   
CUI_Noodle Dishes         31875.0   0.711966   4.537359   0.0   0.0   0.0   
CUI_OTHER                 31875.0   3.001137   9.770104   0.0   0.0   0.0   
CUI_Street Food / Snacks  31875.0   3.910391  15.548367   0.0   0.0   0.0   
CUI_Thai                  31875.0    0.84204   4.433918   0.0   0.0   0.0   
DOW_0                     31875.0   0.555922   1.013634   0.0   0.0   0.0   
DOW_1                     31875.0   0.567561    1.04419   0.0   0.0   0.0   
DOW_2                     31875.0   0.590933     1.0459   0.0   0.0   0.0   
DOW_3                     31875.0   0.619388   1.069558   0.0   0.0   0.0   
DOW_4                     31875.0   0.677867   1.088284   0.0   0.0   0.0   
DOW_5                     31875.0   0.652957   1.069802   0.0   0.0   0.0   
DOW_6                     31875.0   0.704408   1.167592   0.0   0.0   0.0   
HR_0                      30711.0        0.0        0.0   0.0   0.0   0.0   
HR_1                      31875.0   0.053835   0.317031   0.0   0.0   0.0   
HR_2                      31875.0   0.063153    0.35149   0.0   0.0   0.0   
HR_3                      31875.0   0.118306   0.497117   0.0   0.0   0.0   
HR_4                      31875.0    0.10171   0.437549   0.0   0.0   0.0   
HR_5                      31875.0   0.081914   0.358701   0.0   0.0   0.0   
HR_6                      31875.0   0.069616   0.329307   0.0   0.0   0.0   
HR_7                      31875.0   0.076675   0.377349   0.0   0.0   0.0   
HR_8                      31875.0   0.131859   0.635603   0.0   0.0   0.0   
HR_9                      31875.0   0.233945   0.725015   0.0   0.0   0.0   
HR_10                     31875.0   0.329663   0.891312   0.0   0.0   0.0   
HR_11                     31875.0   0.378259   0.960119   0.0   0.0   0.0   
HR_12                     31875.0   0.314259   0.842625   0.0   0.0   0.0   
HR_13                     31875.0   0.236424   0.637513   0.0   0.0   0.0   
HR_14                     31875.0   0.215686   0.599098   0.0   0.0   0.0   
HR_15                     31875.0   0.277114   0.738282   0.0   0.0   0.0   
HR_16                     31875.0    0.35658   0.874598   0.0   0.0   0.0   
HR_17                     31875.0   0.391059   0.943873   0.0   0.0   0.0   
HR_18                     31875.0   0.337098   0.894106   0.0   0.0   0.0   
HR_19                     31875.0   0.245678   0.795433   0.0   0.0   0.0   
HR_20                     31875.0   0.142871   0.586641   0.0   0.0   0.0   
HR_21                     31875.0   0.071184   0.348604   0.0   0.0   0.0   
HR_22                     31875.0   0.048282   0.298324   0.0   0.0   0.0   
HR_23                   

**After analyse that variables, we can conclude that:** <br>  
  1. customer age:
     + average customer is 27.5
     + youngest costumer=15y // oldest customer = 80y
     + 75% of customers are 31yo or younger
     + oldest customer= 80yo (outlier??)
     + most of customers are young, there are a few older individuals in the dataset <br><br>  
  2. vendor count:
     + entries with 0 (needs further exploration)
     + average vendor count is 3
     + 75% of unique vendors customers have ordered from is 4
     + max vendor count is 41 (outlier??)
     + most customers ordered from few vendors, but there are customers with much higher count <br><br>  

  3. product_count:
     + entries with 0 products (no products purchased)
     + max product count is 269 (outlier)
     + most product count is low  <br><br>  

  4. is_chain:
     + relative small amount of orders made in chain restaurants
     + max nº orders made in chain is 83
     + most count of orders made in chain restaurants is low, but there are customers with a very high number of orders in chain (outlier??) <br><br>  
  
  5. first_order:
     + on average customers place their first order 28 days after joining the app
     + st dev = 24.1 suggests a wide spread in values of first orders. (significant variability in the amount of time customers take to make their first order)
     + min = 0 (customers who did their first order today)
     + max = 90 (outlier??)
     + 75% of customers placed their first order 45 days after entering the database <br><br>  

  6. last order:
     + st dev= 23 suggests a wide spread in values of first orders. (significant variability in the amount of time customers take to make their first order)
     + min = 0 (can indicate the customer placed their first order on the first day they joined the dataset/didnt do an order yet. Compare with min first order.) <br><br>  

  7. cuisines types:
     + average may not be a good comparison measure due to the presence of extreme values. mean may be low because high spenders are a low fraction of the total dataset.
     + their spending don't increase the overall avg significance
     + 75% of customers spend nothing or a significant small amount (Asian & American cuisines) in all cuisine types, the maximum values of these cuisines
     + are significantly high (specially for American (280) and Asian (896) cuisines which indicate the presence of potential outliers who frequently order or spend heavily <br><br>  

  8. DW_0-DW_6:
     + 75% of customers ordered 1 time in that day
     + the max indicate the presence of outliers and suggests the peak
     + most customers only ordered 1 time in each day of the week. <br><br>  

  9. hours of the day
      + HR_0 : no activity at midnight
      + HR_1-23: 75% of customers placed no order in these hours. Max can be outlier or unusual behavior

#### <span style="color:salmon"> 1.5 Categorical variables </span>
Categorical variables represent characteristics or qualities that group data into distinct categories or labels.

In [21]:
df.describe(include="object").T

count unique         top   freq
customer_id      31875  31875  fd40d3b0e0      1
customer_region  31433      8        8670   9761
last_promo       15131      3    DELIVERY   6282
payment_method   31875      3        CARD  20153

**After analysing that variables, we can conclude:** 
  1. Customer_region:
     + Customers are from 8 different regions
     + Most customers are located in region 8670
     + Missing values <br><br>  

  2. last_promo:
     + There are promotions in 3 different categories
     + Most customers use promotions in the delivery category
     + Missing values <br><br>  

  3. payment_method:
     + There are 3 different payment methods used by customers
     + Most customers use Card as their preferred payment method <br><br>  

Then, we checked if there was stange values in this variables.

In [22]:
# Different regions where customers are located
# No strange value found
df["customer_region"][df["customer_region"].notna()].unique()

array(['2360', '8670', '4660', '2440', '4140', '2490', '8370', '8550'],
      dtype=object)

In [23]:
# Different types of discounts used by customers
# No strange value found
df["last_promo"][df["last_promo"].notna()].unique()

array(['DELIVERY', 'DISCOUNT', 'FREEBIE'], dtype=object)

In [24]:
# Different types of payment methods used by customers
# No strange value found
df["payment_method"].unique()

array(['DIGI', 'CASH', 'CARD'], dtype=object)

After analysing the categorical variables, we realise that there isn’t strange data.

In [25]:
df = df.set_index("customer_id")

In [26]:
df_explore = pd.DataFrame(df)

# Save to CSV
df_explore.to_csv('../dataset/df_explore.csv', index=False)